File: tests/fixtures/corpus-full/typescript/services.ts
import { User, Product, Order, CreateUserDto } from "./models";
import { Database } from "./db";
import { AuthManager } from "./auth";

export class UserService {
  constructor(private db: Database, private auth: AuthManager) {}

  async register(dto: CreateUserDto): Promise<User> {
    const user: User = {
      id: Date.now(),
      username: dto.username,
      email: dto.email,
      createdAt: new Date(),
      isActive: true,
    };
    await this.db.saveUser(user);
    return user;
  }

  async getProfile(token: string): Promise<User | undefined> {
    return this.auth.getUser(token);
  }

  async deactivate(token: string): Promise<boolean> {
    const user = await this.auth.getUser(token);
    if (!user) return false;
    user.isActive = false;
    await this.db.saveUser(user);
    return true;
  }
}

export class ProductService {
  constructor(private db: Database) {}

  async getById(id: number): Promise<Product | undefined> {
    return this.db.findProduct(id);
  }

  async restock(id: number, qty: number): Promise<boolean> {
    const p = await this.db.findProduct(id);
    if (!p) return false;
    p.stock += qty;
    await this.db.saveProduct(p);
    return true;
  }
}

export class OrderService {
  constructor(private db: Database, private productService: ProductService) {}

  async create(userId: number): Promise<Order> {
    const order: Order = { id: Date.now(), userId, items: [], total: 0, status: "pending" };
    await this.db.saveOrder(order);
    return order;
  }

  async addItem(orderId: number, productId: number, qty: number): Promise<boolean> {
    const [order, product] = await Promise.all([
      this.db.findOrder(orderId),
      this.productService.getById(productId),
    ]);
    if (!order || !product) return false;
    order.items.push({ productId, quantity: qty, price: product.price });
    order.total += product.price * qty;
    await this.db.saveOrder(order);
    return true;
  }
}

File: tests/fixtures/corpus-full/go/services.go
package main

import "time"

type UserService struct {
	db   *Database
	auth *AuthManager
}

func NewUserService(db *Database, auth *AuthManager) *UserService {
	return &UserService{db: db, auth: auth}
}

func (s *UserService) Register(username, email, password string) *User {
	user := &User{
		ID:        int(time.Now().UnixNano()),
		Username:  username,
		Email:     email,
		CreatedAt: time.Now(),
		IsActive:  true,
	}
	s.db.SaveUser(user)
	return user
}

func (s *UserService) GetProfile(token string) (*User, bool) {
	return s.auth.GetUser(token)
}

type ProductService struct {
	db *Database
}

func NewProductService(db *Database) *ProductService {
	return &ProductService{db: db}
}

func (s *ProductService) GetByID(id int) (*Product, bool) {
	return s.db.FindProduct(id)
}

func (s *ProductService) Restock(id, qty int) bool {
	p, ok := s.db.FindProduct(id)
	if !ok {
		return false
	}
	p.Stock += qty
	s.db.SaveProduct(p)
	return true
}

type OrderService struct {
	db         *Database
	productSvc *ProductService
}

func NewOrderService(db *Database, productSvc *ProductService) *OrderService {
	return &OrderService{db: db, productSvc: productSvc}
}

func (s *OrderService) Create(userID int) *Order {
	order := &Order{
		ID:     int(time.Now().UnixNano()),
		UserID: userID,
		Status: "pending",
	}
	s.db.SaveOrder(order)
	return order
}
